lib/omniauth/strategies/oauth2.rb

```diff
@@ -49,7 +50,8 @@ def request_phase
       end
 
       def authorize_params
-        options.authorize_params[:state] = SecureRandom.hex(24)
+        state_length = options[:state_length]
+        options.authorize_params[:state] = SecureRandom.hex(state_length / 2 + 1)[0...state_length]
```

Comment:
maybe will be better to use

```
params = options.authorize_params.merge(options_for("authorize"))
```

and update 

```
       def options_for(option)
         hash = {}
         options.send(:"#{option}_options").select { |key| options[key] }.each do |key|
           hash[key.to_sym] = options[key].respond_to?(:call) ? options[key].call(env) : options[key]
         end
         hash
       end
```

so, allow to use a `Proc` for generation of `state`